pkg/flow/internal/worker/worker_pool.go

```diff
@@ -78,19 +85,30 @@ func (w *shardedWorkerPool) SubmitWithKey(key string, f func()) error {
 		f()
 	}
 	queue := w.workQueues[w.workerFor(key)]
+	deadline := time.Now().Add(timeout)
 
-	w.lock.Lock()
-	defer w.lock.Unlock()
-	if _, exists := w.set[key]; exists {
-		return nil // only queue one job for given key
-	}
-
-	select {
-	case queue <- wrapped:
-		w.set[key] = struct{}{}
-		return nil
-	default:
-		return fmt.Errorf("worker queue is full")
+	for {
```

Comment:
We need a loop where we occasionally release the `w.lock`, because otherwise we could block the execution of the tasks themselves and lead to a deadlock.